marlowe-apps/src/Language/Marlowe/Runtime/App/Transact.hs

```diff
@@ -170,9 +181,36 @@ transactWithEvents event config key request =
             $ \case
               TxInfo{} -> pure ()
               response -> unexpected response
+          when (confirmSeconds > 0)
+            . withSubEvent subEvent (DynamicEventSelector "WaitAfterConfirm")
+            . const
+            $ liftIO . threadDelay . (confirmSeconds *)
+            =<< randomRIO (1_000_000, 2_000_000)
           pure contractId
 
 
+retry
+  :: T.Text
+  -> Event App r DynamicEventSelector f
+  -> [Int]
+  -> (Event App r DynamicEventSelector DynamicField -> App a)
+  -> App a
+retry name event [] action = withSubEvent event (DynamicEventSelector name) action
+retry name event (delay : delays) action =
+  withSubEvent event (DynamicEventSelector name)
+    $ \subEvent ->
+      ExceptT
+        $ runExceptT (action subEvent)
+        >>= \case
+          Right result -> pure $ Right result
+          Left message -> runExceptT
+                            $ do
+                              addField subEvent $ ("failedAttempt" :: T.Text) ≔ message
+                              addField subEvent $ ("waitForRetry" :: T.Text) ≔ delay
+                              liftIO . threadDelay $ delay * 1_000_000
+                              retry name event delays action
```

Comment:
Do you really want the events to nest like this? This recursive call will enter a new `withSubEvent` call without exiting the previous one. This means that the duration of each retry event will include the duration of the subsequent retries as well.

Reply:
I believe so. I'd like to measure how long the whole Contract lifecycle takes, and correlate retries etc. This has been useful for analyzing logs with mongo. I might change my opinion once we switch over to proper log-analysis tools. I plan to set up a meeting with @shlevy to hear his perspective on how I could better use `eventuolly`.

Reply:
I.e., nesting events was easier than doing left joins on event records.